[security] Multi-agent vulnerability audit @ db602b4 — 6 hardening items (no critical/high findings)

A fresh multi-agent security audit was run across the dev server, HTTP/JSON bridge, Lua sandbox surface, runtime + parsers + input, and the build/CI/release pipeline. Each finding below was verified against source at `db602b4`.

**Headline:** no critical or high-severity issues. The hot surfaces that had earned dedicated hardening passes (`#162`, `#136`, `#168`, `#183`, `#184`, `#185`) verify cleanly — see "What was checked and found clean" at the bottom. Six low/info hardening items remain, each scoped tightly enough to file as its own issue if the maintainer prefers that split.

---

## Findings

### F1 · CI · GitHub-official actions float on a major-version tag
**Severity:** low · **Affected:** `.github/workflows/`

Third-party actions are SHA-pinned (good), but the GitHub-official ones still float:

```
.github/workflows/test.yml:14      uses: actions/checkout@v6
.github/workflows/test.yml:63      uses: actions/checkout@v6
.github/workflows/release.yml:16   uses: actions/checkout@v6
.github/workflows/release.yml:111  uses: actions/upload-artifact@v7
.github/workflows/release.yml:146  uses: actions/upload-artifact@v7
.github/workflows/release.yml:161  uses: actions/checkout@v6
.github/workflows/release.yml:164  uses: actions/download-artifact@v8
```

If `actions/*` is compromised upstream, a `@vN` retag would silently pull the bad version into the next CI run — including `release.yml`, which uploads tarballs to a release. The mitigation that exists everywhere else in the workflow (commit-SHA pin) just isn't applied here.

**Fix:** pin each `actions/*` step to a commit SHA, same convention as `laytan/setup-odin@8536d8…` and `softprops/action-gh-release@b4309332…`. Dependabot will keep them current.

---

### F2 · Bridge · JSON decoder has no `lua_checkstack` on recursive descent
**Severity:** low · **Affected:** `src/redin/bridge/json.odin`

`json_decode_value_at` recurses into nested objects/arrays up to `MAX_JSON_DEPTH = 128`. Worst case it pushes a few slots per frame — comfortably under Lua's default 8000-slot stack, so today's binaries are fine. But there is **no `lua_checkstack` anywhere in the bridge package** (verified by `grep`), so the safety margin is implicit: any future change that raises `MAX_JSON_DEPTH`, switches to a thread with a non-default stack, or starts allocating more slots per level can quietly turn a deep payload into a clean Lua abort instead of a `LUA_ERRRUN`.

**Fix:** at the top of `json_decode_value_at` (and the matching encoder recursion if it pushes), call `lua_checkstack(L, 8)` and surface a Lua error string on failure. Defense-in-depth — costs nothing on the happy path.

---

### F3 · Test harness · Unquoted expansion of `*_app.flags` sidecar
**Severity:** low · **Affected:** `test/ui/run-all.sh:111`

```bash
flags_file="$SCRIPT_DIR/${app_name}_app.flags"
extra_flags=()
if [ -f "$flags_file" ]; then
  # shellcheck disable=SC2207
  extra_flags=( $(cat "$flags_file") )           # word-splits + globs
fi
…
"${LAUNCHER[@]}" "$BINARY" "${extra_flags[@]}" "$app_file" &
```

The `shellcheck disable=SC2207` is acknowledging the issue. The contents of a sidecar `.flags` file are word-split *and* glob-expanded by the shell before reaching the binary's argv. Practical risk today is low — the sidecars live under `test/ui/` in the repo, so an attacker would need to land a PR that adds one. But CI runs this script, so a `*_app.flags` containing `* ; curl evil…` would execute on the runner the moment a job picks it up. It's a quiet escalation path that closes for a couple of lines.

**Fix:** read line-by-line instead, e.g.

```bash
extra_flags=()
while IFS= read -r line; do
  [ -n "$line" ] && extra_flags+=( "$line" )
done < "$flags_file"
```

…and either disallow whitespace in a single flag or document that one-flag-per-line is the contract.

---

### F4 · Runtime · No guard against unbounded `:dispatch` recursion or `:dispatch-later` queue growth
**Severity:** info · **Affected:** `src/runtime/effect.fnl`, `src/runtime/dataflow.fnl`

`effect.execute` dispatches synchronously without a depth counter; `:dispatch-later` appends to `timer-queue` without a size cap. A handler that self-dispatches (`{:db db :dispatch [:loop]}`) blows the Lua stack; one that schedules a `dispatch-later` every frame grows `timer-queue` unboundedly.

These are **not security boundaries in redin's threat model** — the app author controls their handlers. But both are easy app-developer footguns that would crash the host with no useful diagnostic. A `MAX_DISPATCH_DEPTH = 64` with a `redin.log` warning on overflow, and a `MAX_TIMER_QUEUE = 10_000` with a drop+warn, would catch these during development without changing the trust model.

**Fix:** add depth-counter / queue-cap with a diagnostic, gated behind a config knob if you want strictness opt-in.

---

### F5 · Runtime · Hot-reload reads watched files without atomic-write protection
**Severity:** info · **Affected:** `src/redin/bridge/hotreload.odin`

Symlink-swap is already defended (`os.lstat`, `#162 L1`). What remains is the editor-saves-non-atomically case: an `mtime`-triggered re-require can land on a half-written buffer, surfacing as a Fennel syntax error and pinning the app on the broken module until the next save. Not a security boundary — the threat is "your own editor + your own filesystem" — but worth a note because the surrounding hardening makes a reader assume it's been thought through.

**Fix:** on parse/load failure, requeue one short retry (e.g. 50ms later) before surfacing the error. Or document the expectation that editors should write atomically (`vim`'s default, `editor:writeAtomic` in VS Code).

---

### F6 · Bridge · Full Lua stdlib (`os`, `io`, `debug`, `package`) exposed to app code, undocumented
**Severity:** info · **Affected:** `src/redin/bridge/bridge.odin:73` (`luaL_openlibs`)

This is **intentional and correct** under redin's single-principal trust model — the host already gives apps `redin.shell` and `redin.http`, so withholding `os.execute` would be theatre. But the docs don't say so explicitly. A downstream embedder reading `docs/reference/native-bridge.md` could plausibly assume that "Fennel/Lua app" means "sandboxed Lua app" and design their distribution model on that assumption.

**Fix:** add one paragraph to `docs/reference/native-bridge.md` (or wherever you describe the app loading model) stating that app `.fnl` / `.lua` files run with full Lua stdlib access, are trusted as the principal, and that redin is not a sandbox for untrusted code. Anchor any future "embed redin to run third-party widgets" idea to this constraint.

---

## What was checked and found clean

These are not findings — included so the diff between this audit and any prior one is legible:

- **Dev server auth path:** Host-header check (case-insensitive `equal_fold` against `127.0.0.1:<port>` / `localhost:<port>`) runs before auth; bearer-token comparison is constant-time; OPTIONS doesn't bypass anything because it returns 405. (`devserver.odin:638`, `:658`, `:810`)
- **Token file lifecycle:** `O_EXCL|O_NOFOLLOW`, mode 0600, EEXIST handled via `lstat` + cleanup, removed in `devserver_destroy` *and* via a signal handler with `SA_RESETHAND`. Ordering on shutdown is correct — files removed only after all threads exit. (`devserver.odin:242–264`, `:307–314`, `:426–427`, `:197–204`)
- **SSRF:** Whitelist-deny model with access-class IP enforcement; host is resolved once and the **resolved endpoint** is dialled, defeating DNS-rebind TOCTOU between getaddrinfo and connect. Loopback, RFC1918, link-local v4/v6, CGNAT, ULA, and cloud-metadata IPs are blocked unless explicitly whitelisted. (`http_client.odin:504–532`, `api.odin:423–565`; tests in `http_access_test.odin`)
- **Header / URL CRLF injection:** `header_safe()` (`http_client.odin:35–71`) rejects `\r`, `\n`, NUL in URL and every header k/v; applied at `:498` and `:559`.
- **HTTP body / timeouts:** 16 MiB body cap, 30s default timeout, per-recv deadline at socket level. (`http_client.odin:15–32`, `:602`)
- **JSON parser bounds:** `MAX_JSON_DEPTH = 128`, `MAX_JSON_ARRAY_LEN = 1_000_000` (i32-wrap guard), surrogate-pair validation per RFC 8259, cycle detection on the recursion path. (`json.odin:86`, `:401`, `:303–316`, `:88–97`)
- **Request parsing:** 1 MiB body cap, 10s per-request deadline, 2s per-recv timeout, `Content-Length` digits capped, 16-worker pool / 256-queue. (`devserver.odin:51–58`, `:452`, `:461`, `:497`, `:830`)
- **Screenshot:** PNG encoded in-memory via `rl.ExportImageToMemory`, no temp file, 16 MP pixel cap. (`devserver.odin:2312–2337`, `:2301`)
- **Synthesised input:** `/click`, `/input/mouse/move`, `/input/scroll`, `/input/key` all reject NaN/Inf and out-of-screen coords; key endpoint maps from an allowlist. (`devserver.odin:1839`, `:1926`, `:2082`, `:2164`)
- **Shell host function:** env allowlist (empty by default — deny-all), 64 MiB stdin cap, 16 MiB stdout+stderr cap, type-checked argv. (`shell.odin:12,14,198–206`, `api.odin:705–799`)
- **Build scripts:** all variable expansions quoted; `mktemp -d` with EXIT-trap cleanup; Babashka pinned by version *and* SHA-256 checksum with `sha256sum -c` before extraction.
- **Path-step encoding for selection (`#180`)** and **shell process-group kill (`#171`)** were already filed; not re-raised.
- **Release tarball signing (`#158`)** remains the right tracking issue for cosign keyless — not duplicated here.

---

## Threat-model note

Severity throughout reflects redin's stated threat model: a **localhost + per-run-token** dev server and **trusted app Fennel/Lua** (which already has `shell`, `http`, and full stdlib). Findings that would be high-severity for an Internet-facing service or a sandbox-for-untrusted-code framework are correctly low/info here.

---

_Found via an automated multi-agent code audit; each finding was verified against source at `db602b4`._

--- a/.github/workflows/release.yml
+++ b/.github/workflows/release.yml
@@ -13,7 +13,7 @@ jobs:
     runs-on: ubuntu-latest
 
     steps:
-      - uses: actions/checkout@v6
+      - uses: actions/checkout@df4cb1c069e1874edd31b4311f1884172cec0e10  # v6.0.3
         with:
           submodules: recursive
 
@@ -108,7 +108,7 @@ jobs:
         run: bash scripts/smoke-native.sh "redin-${VERSION}-linux-amd64.tar.gz"
 
       - name: Upload artifact
-        uses: actions/upload-artifact@v7
+        uses: actions/upload-artifact@043fb46d1a93c77aae656e7c1c64a875d1fc6a0a  # v7.0.1
         with:
           name: redin-${{ steps.version.outputs.tag }}-linux-amd64
           path: |
@@ -143,7 +143,7 @@ jobs:
             > "redin-${VERSION}-agent-linux-amd64.tar.gz.sha256"
 
       - name: Upload agent artifact
-        uses: actions/upload-artifact@v7
+        uses: actions/upload-artifact@043fb46d1a93c77aae656e7c1c64a875d1fc6a0a  # v7.0.1
         with:
           name: redin-${{ steps.version.outputs.tag }}-agent-linux-amd64
           path: |
@@ -158,10 +158,10 @@ jobs:
       contents: write
 
     steps:
-      - uses: actions/checkout@v6
+      - uses: actions/checkout@df4cb1c069e1874edd31b4311f1884172cec0e10  # v6.0.3
 
       - name: Download artifacts
-        uses: actions/download-artifact@v8
+        uses: actions/download-artifact@3e5f45b2cfb9172054b4087a40e8e0b5a5461e7c  # v8.0.1
         with:
           path: artifacts
 

--- a/.github/workflows/test.yml
+++ b/.github/workflows/test.yml
@@ -11,7 +11,7 @@ jobs:
     runs-on: ubuntu-latest
 
     steps:
-      - uses: actions/checkout@v6
+      - uses: actions/checkout@df4cb1c069e1874edd31b4311f1884172cec0e10  # v6.0.3
         with:
           submodules: recursive
 
@@ -60,7 +60,7 @@ jobs:
     runs-on: ubuntu-latest
 
     steps:
-      - uses: actions/checkout@v6
+      - uses: actions/checkout@df4cb1c069e1874edd31b4311f1884172cec0e10  # v6.0.3
         with:
           submodules: recursive
 

--- a/docs/app-api.md
+++ b/docs/app-api.md
@@ -350,6 +350,10 @@ Dispatch another event immediately.
  :dispatch [:event/notify "Saved"]}
 ```
 
+Dispatches run synchronously, so a self-dispatching chain is capped at depth
+**64** — past it the event is dropped with a warning instead of overflowing
+the host stack. See [reference/effects.md](reference/effects.md#built-in-dispatch).
+
 #### `:dispatch-later`
 
 Schedule event dispatch after a delay.
@@ -365,7 +369,7 @@ Schedule event dispatch after a delay.
                   {:ms 5000 :dispatch [:session/timeout]}]}
 ```
 
-Each entry: `{:ms N :dispatch event-vector}`. Timer fires when `poll-timers` is called with `now >= start + ms`. The host calls `poll-timers` once per frame.
+Each entry: `{:ms N :dispatch event-vector}`. Timer fires when `poll-timers` is called with `now >= start + ms`. The host calls `poll-timers` once per frame. The pending-timer queue is capped at **10,000** entries; beyond it new timers are dropped with a warning.
 
 #### `:log`
 

--- a/docs/reference/effects.md
+++ b/docs/reference/effects.md
@@ -47,6 +47,13 @@ Dispatch another event immediately.
  :dispatch [:event/notify "Saved"]}
 ```
 
+A `:dispatch` runs synchronously, so a handler that dispatches an event whose
+handler dispatches back forms a recursive chain. That chain is capped at depth
+**64**: past it the event is dropped with a `Warning: dispatch depth exceeded`
+log rather than overflowing the host stack. The cap is a footgun guard for
+unbounded self-dispatch, not a normal control-flow limit — real chains are
+far shallower.
+
 ---
 
 ## Built-in: `:dispatch-later`
@@ -63,6 +70,11 @@ Schedule event dispatch after a delay. Accepts a single timer map or a sequence 
 ```
 
 Each entry: `{:ms N :dispatch event-vector}`. The timer fires when `poll-timers` is called with `now >= start + ms`. The host calls `poll-timers` once per frame.
+
+The pending-timer queue is capped at **10,000** entries. A handler that
+schedules a timer every frame without them firing would otherwise grow the
+queue without bound; past the cap, new `:dispatch-later` timers are dropped
+with a `Warning: timer queue at cap` log. No legitimate use approaches this.
 
 ---
 

--- a/docs/reference/native-bridge.md
+++ b/docs/reference/native-bridge.md
@@ -8,6 +8,31 @@ Import from your `app.odin`:
 import "./.redin/src/redin/bridge"
 ```
 
+## Trust model
+
+**redin is not a sandbox.** Your app's `.fnl` / `.lua` files are the trusted
+principal, not untrusted content. They load with the **full Lua standard
+library** — including `os` (`os.execute`, `os.getenv`, `os.remove`), `io`,
+`debug`, and `package` — and on top of that the host hands them `redin.shell`
+and `redin.http`. App code can therefore run arbitrary commands and read or
+write the filesystem with the privileges of the process. This is intentional:
+the app *is* the program, so withholding `os.execute` while shipping
+`redin.shell` would be theatre.
+
+The deny-by-default effect policies below (`set_http_whitelist`,
+`set_shell_env_allowlist`) and the SSRF/CRLF hardening on `redin.http` are
+defense-in-depth for the *host-configured* surface — they bound what a bug or
+a compromised remote endpoint can reach. They are **not** a boundary that
+contains hostile app code: code with `os`/`io`/`package` can ignore the
+framework entirely. Treat the `.fnl`/`.lua` you load with the same trust you'd
+give a native plugin compiled into the binary.
+
+Concretely: **do not** use redin to run third-party `.fnl`/`.lua` widgets,
+user-submitted scripts, or any code you would not run as a shell command. If
+you ever need to host untrusted UI code, the isolation must come from outside
+redin (a separate process, an OS sandbox, a container) — there is no
+in-framework setting that makes loading untrusted scripts safe.
+
 ## Cfunc registration
 
 ### `bridge.register_cfunc(name: cstring, fn: proc(L: ^Lua_State) -> i32)`

--- a/src/redin/bridge/bridge.odin
+++ b/src/redin/bridge/bridge.odin
@@ -192,10 +192,17 @@ is_shutdown_requested :: proc(b: ^Bridge) -> bool {
 
 check_hotreload :: proc(b: ^Bridge) {
 	when !REDIN_DEV do return
-	if hotreload_check(&b.hot_reload) {
-		hotreload_execute(b)
-		b.frame_changed = true
+	trigger := hotreload_poll(&b.hot_reload)
+	if trigger == .None do return
+	ok := hotreload_execute(b)
+	// F5: a fresh-change reload that failed is likely a non-atomic editor
+	// save caught mid-write — arm one short retry. A retry that also fails
+	// (or any success) leaves the retry disarmed, so a genuinely broken file
+	// settles after two attempts instead of retrying forever.
+	if hotreload_should_arm_retry(trigger, ok) {
+		hotreload_arm_retry(&b.hot_reload)
 	}
+	b.frame_changed = true
 }
 
 // Walk the flat tree once: for each draggable with handle_off, ensure at

--- a/src/redin/bridge/hotreload.odin
+++ b/src/redin/bridge/hotreload.odin
@@ -4,11 +4,32 @@ import "core:fmt"
 import "core:os"
 import "core:time"
 
+// F5 (#204): how long after a failed reload to make one more attempt.
+// An editor that saves non-atomically (write-in-place rather than
+// write-temp-then-rename) can update a watched file's mtime while the
+// buffer is still half-written, so the mtime-triggered re-require lands on
+// a truncated module and fails with a Fennel syntax error. A single short
+// retry gives the editor time to finish the write; 50ms is far longer than
+// any in-place save and short enough to feel instant.
+HOTRELOAD_RETRY_MS :: 50.0
+
+Hotreload_Trigger :: enum {
+	None,    // nothing to do this frame
+	Changed, // a watched file's mtime advanced
+	Retry,   // a previously-armed one-shot retry came due
+}
+
 Hot_Reload :: struct {
 	watch_paths:    [dynamic]string,
 	last_mtimes:    map[string]i64,
 	frame_counter:  int,
 	check_interval: int,
+	// F5: one-shot retry after a failed reload. retry_armed is true between
+	// arming the retry and consuming it; retry_at marks when it was armed.
+	// Because it is armed only on a fresh-change failure and consumed once,
+	// a genuinely broken file fails at most twice, never in a loop.
+	retry_armed:    bool,
+	retry_at:       time.Tick,
 }
 
 hotreload_init :: proc(hr: ^Hot_Reload, source_tree: bool) {
@@ -38,9 +59,36 @@ hotreload_destroy :: proc(hr: ^Hot_Reload) {
 	delete(hr.last_mtimes)
 }
 
-hotreload_check :: proc(hr: ^Hot_Reload) -> bool {
+// Pure predicate: given a retry is armed and `elapsed_ms` have passed since
+// it was armed, is it due to fire? Factored out so the timing rule is unit-
+// testable without a real clock.
+hotreload_retry_due :: proc(armed: bool, elapsed_ms: f64) -> bool {
+	return armed && elapsed_ms >= HOTRELOAD_RETRY_MS
+}
+
+// Pure predicate: after an execute attempt, should a one-shot retry be armed?
+// Only when a *fresh* mtime change (not an already-consumed retry) failed —
+// this is what bounds a broken file to two attempts instead of a loop.
+hotreload_should_arm_retry :: proc(trigger: Hotreload_Trigger, reload_ok: bool) -> bool {
+	return trigger == .Changed && !reload_ok
+}
+
+hotreload_poll :: proc(hr: ^Hot_Reload) -> Hotreload_Trigger {
+	// A pending retry fires on its own time-based deadline, independent of
+	// the mtime-poll throttle, and is consumed here so it fires at most once.
+	// While a retry is still pending we hold off on mtime polling so the two
+	// paths can't interleave.
+	if hr.retry_armed {
+		elapsed := time.duration_milliseconds(time.tick_since(hr.retry_at))
+		if hotreload_retry_due(hr.retry_armed, elapsed) {
+			hr.retry_armed = false
+			return .Retry
+		}
+		return .None
+	}
+
 	hr.frame_counter += 1
-	if hr.frame_counter < hr.check_interval do return false
+	if hr.frame_counter < hr.check_interval do return .None
 	hr.frame_counter = 0
 
 	changed := false
@@ -51,10 +99,17 @@ hotreload_check :: proc(hr: ^Hot_Reload) -> bool {
 		}
 		hr.last_mtimes[path] = mtime
 	}
-	return changed
+	return changed ? .Changed : .None
 }
 
-hotreload_execute :: proc(b: ^Bridge) {
+// Arm a single retry to fire ~HOTRELOAD_RETRY_MS later. retry_at records
+// the moment of failure; hotreload_poll fires once that long has elapsed.
+hotreload_arm_retry :: proc(hr: ^Hot_Reload) {
+	hr.retry_armed = true
+	hr.retry_at = time.tick_now()
+}
+
+hotreload_execute :: proc(b: ^Bridge) -> (ok: bool) {
 	code := `
 		package.loaded["dataflow"] = nil
 		package.loaded["effect"]   = nil
@@ -76,9 +131,10 @@ hotreload_execute :: proc(b: ^Bridge) {
 		msg := lua_tostring_raw(b.L, -1)
 		fmt.eprintfln("Hot reload error: %s", msg)
 		lua_pop(b.L, 1)
-	} else {
-		fmt.println("Hot reload: runtime reloaded")
+		return false
 	}
+	fmt.println("Hot reload: runtime reloaded")
+	return true
 }
 
 @(private = "file")

--- a/src/redin/bridge/hotreload_retry_test.odin
+++ b/src/redin/bridge/hotreload_retry_test.odin
@@ -1,0 +1,38 @@
+package bridge
+
+// F5 (#204): a hot reload triggered by a non-atomic editor save can land on
+// a half-written file and fail. The watcher arms one short retry so the
+// editor's finished write is picked up automatically. The timing rule and
+// the arm-at-most-once rule are factored into pure predicates so they can be
+// checked without a real clock, Lua state, or filesystem.
+
+import "core:testing"
+
+@(test)
+test_hotreload_retry_due_timing :: proc(t: ^testing.T) {
+	// Not armed: never due, regardless of elapsed time.
+	testing.expect(t, !hotreload_retry_due(false, 0), "disarmed never fires")
+	testing.expect(t, !hotreload_retry_due(false, 1000), "disarmed never fires even much later")
+
+	// Armed but before the deadline: not yet.
+	testing.expect(t, !hotreload_retry_due(true, 0), "armed but no time passed")
+	testing.expect(t, !hotreload_retry_due(true, HOTRELOAD_RETRY_MS - 1), "armed but just under the deadline")
+
+	// Armed and at/after the deadline: due.
+	testing.expect(t, hotreload_retry_due(true, HOTRELOAD_RETRY_MS), "exactly at the deadline fires")
+	testing.expect(t, hotreload_retry_due(true, HOTRELOAD_RETRY_MS + 5), "past the deadline fires")
+}
+
+@(test)
+test_hotreload_should_arm_retry :: proc(t: ^testing.T) {
+	// Only a fresh mtime change that failed arms a retry.
+	testing.expect(t, hotreload_should_arm_retry(.Changed, false), "failed fresh change arms a retry")
+
+	// Everything else leaves it disarmed — crucially, a failed *retry* does
+	// not arm another, bounding a broken file to two attempts.
+	testing.expect(t, !hotreload_should_arm_retry(.Changed, true), "successful change does not arm")
+	testing.expect(t, !hotreload_should_arm_retry(.Retry, false), "failed retry does not re-arm (no loop)")
+	testing.expect(t, !hotreload_should_arm_retry(.Retry, true), "successful retry does not arm")
+	testing.expect(t, !hotreload_should_arm_retry(.None, false), "no trigger never arms")
+	testing.expect(t, !hotreload_should_arm_retry(.None, true), "no trigger never arms")
+}

--- a/src/redin/bridge/json.odin
+++ b/src/redin/bridge/json.odin
@@ -108,6 +108,13 @@ lua_value_to_json_inner :: proc(b: ^strings.Builder, L: ^Lua_State, index: i32, 
 		json_null(b)
 		return
 	}
+	// Same stack guard as the decoder. The encoder keeps a strings.Builder
+	// alive up the call stack, so on exhaustion we degrade to `null` (as the
+	// depth cap does) rather than longjmp out via luaL_error and leak it.
+	if lua_checkstack(L, 8) == 0 {
+		json_null(b)
+		return
+	}
 	abs_idx := index < 0 ? lua_gettop(L) + index + 1 : index
 	t := lua_type(L, abs_idx)
 	switch t {
@@ -229,6 +236,17 @@ json_decode_value :: proc(L: ^Lua_State, s: string, pos: ^int) -> bool {
 @(private = "file")
 json_decode_value_at :: proc(L: ^Lua_State, s: string, pos: ^int, depth: int) -> bool {
 	if depth > MAX_JSON_DEPTH do return false
+	// Defense-in-depth: guarantee room for the slots this level and its
+	// children push (table + key + value, plus headroom) before descending.
+	// MAX_JSON_DEPTH (128) keeps today's worst case far under LuaJIT's
+	// 8000-slot default, so this never trips now; it stops a future raised
+	// cap or a non-default-stack thread from turning a deep payload into an
+	// unchecked overflow instead of a catchable error. No Odin allocation is
+	// live on this recursion path, so the luaL_error longjmp leaks nothing.
+	if lua_checkstack(L, 8) == 0 {
+		luaL_error(L, "JSON nesting exhausts Lua stack")
+		return false
+	}
 	json_skip_ws(s, pos)
 	if pos^ >= len(s) do return false
 	c := s[pos^]

--- a/src/redin/bridge/json_checkstack_test.odin
+++ b/src/redin/bridge/json_checkstack_test.odin
@@ -1,0 +1,65 @@
+package bridge
+
+// F2 (#204): the JSON encoder/decoder recurse per nesting level, pushing a
+// few Lua slots each. MAX_JSON_DEPTH (128) keeps the worst case far under
+// LuaJIT's 8000-slot default, but the recursion now calls lua_checkstack
+// before descending so a future raised cap or a non-default-stack thread
+// fails cleanly instead of overflowing the stack unchecked.
+//
+// The exhaustion path itself can't be provoked deterministically at depth
+// 128, so these tests pin the two things that can regress: the binding is
+// wired (checkstack reports headroom), and the guard leaves the happy path
+// — a deep-but-valid payload — decoding correctly while the depth cap still
+// rejects past the limit.
+
+import "core:strings"
+import "core:testing"
+
+@(test)
+test_lua_checkstack_binding_reports_headroom :: proc(t: ^testing.T) {
+	L := luaL_newstate()
+	luaL_openlibs(L)
+	defer lua_close(L)
+
+	// A fresh state has ample room; the guard's happy path must pass.
+	testing.expect(t, lua_checkstack(L, 8) != 0, "checkstack(8) should succeed on a fresh state")
+}
+
+// Build `[`×n + `]`×n: an array nested n levels deep with an empty array
+// at the bottom. The innermost array is decoded by json_decode_value_at at
+// depth n-1, so n stays clear of the depth cap.
+@(private = "file")
+nested_arrays :: proc(n: int) -> string {
+	b := strings.builder_make()
+	for _ in 0 ..< n do strings.write_byte(&b, '[')
+	for _ in 0 ..< n do strings.write_byte(&b, ']')
+	return strings.to_string(b)
+}
+
+@(test)
+test_decode_deep_within_cap_succeeds :: proc(t: ^testing.T) {
+	L := luaL_newstate()
+	luaL_openlibs(L)
+	defer lua_close(L)
+
+	// 100 levels: deep enough to exercise the per-level checkstack guard
+	// many times, comfortably under MAX_JSON_DEPTH (128).
+	s := nested_arrays(100)
+	defer delete(s)
+	pos := 0
+	testing.expect(t, json_decode_value(L, s, &pos), "100-level nesting is within the cap and must decode")
+}
+
+@(test)
+test_decode_over_cap_rejected :: proc(t: ^testing.T) {
+	L := luaL_newstate()
+	luaL_openlibs(L)
+	defer lua_close(L)
+
+	// 200 levels is well past MAX_JSON_DEPTH; the depth guard rejects it
+	// before the stack guard would ever matter.
+	s := nested_arrays(200)
+	defer delete(s)
+	pos := 0
+	testing.expect(t, !json_decode_value(L, s, &pos), "200-level nesting exceeds the depth cap and must be rejected")
+}

--- a/src/redin/bridge/lua_api.odin
+++ b/src/redin/bridge/lua_api.odin
@@ -28,6 +28,10 @@ foreign luajit {
 
 	lua_gettop    :: proc(L: ^Lua_State) -> i32 ---
 	lua_settop    :: proc(L: ^Lua_State, index: i32) ---
+	// Ensures at least `extra` free slots on the Lua stack, growing it if
+	// needed. Returns 0 (false) when the stack cannot grow. Calling a push
+	// without this guard past the available slots is an unchecked overflow.
+	lua_checkstack :: proc(L: ^Lua_State, extra: i32) -> i32 ---
 	lua_pushvalue :: proc(L: ^Lua_State, index: i32) ---
 	lua_remove    :: proc(L: ^Lua_State, index: i32) ---
 	lua_type      :: proc(L: ^Lua_State, index: i32) -> i32 ---

--- a/src/runtime/dataflow.fnl
+++ b/src/runtime/dataflow.fnl
@@ -10,6 +10,16 @@
 (var tracking nil)
 (var effect-handler nil)
 
+;; F4 (#204): cap on synchronous dispatch recursion. A handler that returns
+;; {:db db :dispatch [...]} re-enters dispatch through the effect handler; one
+;; that self-dispatches unconditionally would otherwise recurse until the Lua
+;; stack overflows with no useful diagnostic. Not a security boundary — the
+;; app author controls their handlers — but an easy footgun. 64 is far deeper
+;; than any legitimate dispatch chain.
+(local MAX-DISPATCH-DEPTH 64)
+(set M.MAX_DISPATCH_DEPTH MAX-DISPATCH-DEPTH)
+(var dispatch-depth 0)
+
 ;; ===== Path utilities =====
 
 (fn paths-overlap? [a b]
@@ -122,7 +132,7 @@
 (fn M.reg-handler [key handler-fn]
   (tset handlers key handler-fn))
 
-(fn M.dispatch [event]
+(fn do-dispatch [event]
   (let [key (. event 1)
         handler-fn (. handlers key)]
     (assert handler-fn (.. "No handler registered for: " (tostring key)))
@@ -134,6 +144,23 @@
                    (= (type result) "table") (~= (. result :db) nil))
           (when effect-handler
             (effect-handler result)))))))
+
+(fn M.dispatch [event]
+  (if (>= dispatch-depth MAX-DISPATCH-DEPTH)
+    ;; Runaway synchronous recursion — drop the event with a diagnostic
+    ;; instead of overflowing the Lua stack.
+    (print (.. "Warning: dispatch depth exceeded " MAX-DISPATCH-DEPTH
+               " (runaway :dispatch recursion?); dropping event "
+               (tostring (. event 1))))
+    (do
+      (set dispatch-depth (+ dispatch-depth 1))
+      ;; pcall so the depth counter is restored even when a handler errors;
+      ;; otherwise one errored dispatch leaves the counter elevated and
+      ;; eventually wedges all dispatching. Re-raise with level 0 to preserve
+      ;; the original message (e.g. the "No handler" assert) verbatim.
+      (let [(ok? err) (pcall do-dispatch event)]
+        (set dispatch-depth (- dispatch-depth 1))
+        (when (not ok?) (error err 0))))))
 
 ;; ===== Public API: change detection =====
 
@@ -191,7 +218,8 @@
   (set subscriptions {})
   (set changed-paths [])
   (set tracking nil)
-  (set effect-handler nil))
+  (set effect-handler nil)
+  (set dispatch-depth 0))
 
 ;; ===== Global registration =====
 

--- a/src/runtime/effect.fnl
+++ b/src/runtime/effect.fnl
@@ -3,6 +3,15 @@
 
 (local M {})
 
+;; F4 (#204): upper bound on outstanding dispatch-later timers. A handler
+;; that schedules a timer every frame grows timer-queue without bound and
+;; slowly exhausts memory with no useful diagnostic. Not a security boundary
+;; — the app author controls their handlers — but an easy footgun. At the cap
+;; we drop new timers with a warning rather than grow forever. 10,000 is far
+;; above any legitimate fan-out of pending timers.
+(local MAX-TIMER-QUEUE 10000)
+(set M.MAX_TIMER_QUEUE MAX-TIMER-QUEUE)
+
 (var executors {})
 (var timer-queue [])
 (var pending-http {})
@@ -14,6 +23,16 @@
 
 (fn M.reg-fx [key executor-fn]
   (tset executors key executor-fn))
+
+;; Enqueue a timer unless the queue is at its cap, in which case drop it
+;; with a warning. Centralised so both the single- and list-form of
+;; :dispatch-later share the bound (F4, #204).
+(fn enqueue-timer [at event]
+  (if (< (length timer-queue) MAX-TIMER-QUEUE)
+    (table.insert timer-queue {:at at :event event})
+    (print (.. "Warning: timer queue at cap (" MAX-TIMER-QUEUE
+               "); dropping dispatch-later event "
+               (tostring (and (= (type event) "table") (. event 1)))))))
 
 ;; ===== Execution =====
 
@@ -91,9 +110,9 @@
                   (* (_G.redin.now) 1000)
                   (* (os.clock) 1000))]
         (if (and (= (type params) "table") (. params :ms))
-          (table.insert timer-queue {:at (+ now params.ms) :event params.dispatch})
+          (enqueue-timer (+ now params.ms) params.dispatch)
           (each [_ timer-spec (ipairs params)]
-            (table.insert timer-queue {:at (+ now timer-spec.ms) :event timer-spec.dispatch}))))))
+            (enqueue-timer (+ now timer-spec.ms) timer-spec.dispatch))))))
   (M.reg-fx :http
     (fn [params]
       (set next-http-id (+ next-http-id 1))

--- a/test/lua/test_effect.fnl
+++ b/test/lua/test_effect.fnl
@@ -104,6 +104,46 @@
   (effect.clear-timers)
   (assert (= (effect.pending-timers) 0) "timers cleared"))
 
+;; --- F4 (#204): runaway-growth guards ---
+
+;; A handler-less `:dispatch-later` scheduled every frame grows timer-queue
+;; without bound. The cap drops new timers (with a warning) once the queue
+;; is full rather than letting it consume memory forever.
+(fn t.test-timer-queue-cap-drops-overflow []
+  (setup)
+  (let [cap effect.MAX_TIMER_QUEUE]
+    (assert (and cap (> cap 0)) "MAX_TIMER_QUEUE is exported and positive")
+    (for [_ 1 cap]
+      (effect.execute {:db nil :dispatch-later {:ms 100 :dispatch [:event/x]}}))
+    (assert (= (effect.pending-timers) cap) "queue fills exactly to the cap")
+    ;; One more past the cap must be dropped, not enqueued.
+    (effect.execute {:db nil :dispatch-later {:ms 100 :dispatch [:event/x]}})
+    (assert (= (effect.pending-timers) cap) "over-cap dispatch-later is dropped")))
+
+;; A handler that unconditionally re-dispatches itself recurses
+;; synchronously through effect.execute -> :dispatch -> dataflow.dispatch.
+;; Without a depth cap this overflows the Lua stack; with it, the chain is
+;; cut at MAX_DISPATCH_DEPTH and unwinds cleanly (no error).
+(fn t.test-dispatch-depth-cap-stops-runaway []
+  (setup)
+  (dataflow.init {:n 0})
+  (dataflow.register-globals)
+  (dataflow.set-effect-handler effect.execute)
+  (dataflow.reg-handler :event/loop
+    (fn [db _event]
+      (dataflow.update db :n #(+ (or $1 0) 1))
+      {:db db :dispatch [:event/loop]}))
+  (let [(ok err) (pcall dataflow.dispatch [:event/loop])]
+    (assert ok (.. "runaway dispatch must be capped, not error: " (tostring err))))
+  ;; Exactly MAX_DISPATCH_DEPTH handlers run before the cap drops the next.
+  (assert (= (rawget (dataflow._get-raw-db) :n) dataflow.MAX_DISPATCH_DEPTH)
+          "ran exactly MAX_DISPATCH_DEPTH handlers before hitting the cap")
+  ;; The depth counter must return to baseline so later dispatches aren't
+  ;; wrongly rejected.
+  (dataflow.dispatch [:event/loop])
+  (assert (= (rawget (dataflow._get-raw-db) :n) (* 2 dataflow.MAX_DISPATCH_DEPTH))
+          "depth counter reset between top-level dispatches"))
+
 ;; Regression test for issue #146.
 ;;
 ;; In production the host calls poll-timers with wall-clock ms

--- a/test/ui/run-all.sh
+++ b/test/ui/run-all.sh
@@ -103,12 +103,20 @@ for test_file in "$SCRIPT_DIR"/test_*.bb; do
 
   echo "=== $name ==="
 
-  # Optional sidecar: <app>_app.flags — whitespace-split extra host flags.
+  # Optional sidecar: <app>_app.flags — extra host flags, ONE PER LINE.
+  # Read line-by-line (no word-splitting, no globbing): the file contents
+  # are repo-controlled, but CI runs this script, so an unquoted `$(cat …)`
+  # expansion would let a `*_app.flags` carrying shell metacharacters run on
+  # the runner the moment a job picks it up (#204 F3). One flag per line
+  # means whitespace inside a line is preserved as part of that single arg.
+  # Blank lines are ignored; the `|| [ -n "$line" ]` tail catches a final
+  # line with no trailing newline.
   flags_file="$SCRIPT_DIR/${app_name}_app.flags"
   extra_flags=()
   if [ -f "$flags_file" ]; then
-    # shellcheck disable=SC2207
-    extra_flags=( $(cat "$flags_file") )
+    while IFS= read -r line || [ -n "$line" ]; do
+      [ -n "$line" ] && extra_flags+=( "$line" )
+    done < "$flags_file"
   fi
 
   # Start dev server in background

